TrinityCore/TrinityCore
# Core/Players: Improve trade exchange calculations

**Changes proposed:**

-  When determining whether or not an item can be exchanged in `Player::CanStoreItems`, spread incoming pieces of a stack across multiple pieces of the mock-inventory if necessary.
- Allow merging of incoming items into one free inventory slot.

**Target branch(es):** 3.3.5

- [x] 3.3.5

**Issues addressed:** Closes #24434

**Tests performed:**

Builds successfully.

Test 1: Trade across multiple stacks

1. Place 2 characters together, Character A and Character B.
2. Give the following items to A:
	* 38 pieces of Linen Cloth: `.additem 2589 38`
	* Enough non-Linen Cloth items to fill inventory: `.additem 6265 255`
3. Character A should arrange the Linen Cloth to be in 2 stacks of 19
4. Give the following items to Character B:
	* 2 pieces of Linen Cloth: `.additem 2589 2`
5. Open a trade between Character A and Character B
6. Character B shall attempt to trade their 2 pieces of Linen to Character A

Observed result (Retail Classic): Trade completes without errors
Expected result (3.3.5, broken): Trade shall fail on account of Character A not having enough space
Expected result (branch): Trade shall go through, leaving Character A with 2 stacks of 20 Linen Cloth

This problem was that each inventory-slot calculation in `Player::CanStoreItems` only accepts a full stack merge. It would not accept 2+ partial matches.

Test 2: Multiple incoming stacks competing for free space

1. Place 2 characters together, Character A and Character B.
2. Fill player A's inventory with something like Soul Shards: `.additem 6265 255`
3. Free up one space in player A's inventory.
4. Give Player B 2 Incendite Ore: `.additem 3340 2`
5. Player B should break these 2 pieces into 2 stacks.
6. Open a trade between Character A and Character B
7. Character B shall attempt to trade their 2 incomplete stacks of Incendite Ore to Player A.

Expected result (3.3.5, broken): Trade shall fail on account of Character A not having enough space
Expected result (branch): Trade shall go through, leaving Character A with a stack of 2 Incendite Ore.

Test 3: Free space error check

1. Place 2 characters together, Character A and Character B.
2. Fill player A's inventory with something like Soul Shards: `.additem 6265 255`
3. Free up one space in player A's inventory.
4. Give Player B 12 Incendite Ore: `.additem 3340 12`
5. Player B should break these 12 pieces into 2 stacks of 6.
6. Open a trade between Character A and Character B
7. Character B shall attempt to trade their 2 incomplete stacks of Incendite Ore to Player A.

Expected result (branch): Trade shall fail, as the receiving player has no space

Test 4: Free space error check

1. Place 2 characters together, Character A and Character B.
2. Fill player A's inventory with something like Soul Shards: `.additem 6265 255`
3. Free up two spaces in player A's inventory.
4. Give Player A 8 Incendite Ore: `.additem 3340 8`
5. Give Player B 12 Incendite Ore: `.additem 3340 12`
6. Player B should break these 12 pieces into 2 stacks of 6.
7. Open a trade between Character A and Character B
8. Character B shall attempt to trade their 2 incomplete stacks of Incendite Ore to Player A.

Expected result (branch): Trade shall go through, leaving Character A with 12 pieces of Ore.

This test confirmed that the changes aren't being too generous with allocated space.

**Known issues and TODO list:**
- [x] Fix Test 2. At time of writing, it is failing. This also blocks tests 3 and 4.
